src/main/scala/com/databricks/labs/overwatch/pipeline/SilverTransforms.scala

```diff
@@ -1314,4 +1314,119 @@ trait SilverTransforms extends SparkSessionWrapper {
 
   }
 
+  private def warehouseBase(auditRawDF: DataFrame): DataFrame = {
+    val warehouse_id_gen_w = Window.partitionBy('organization_id, 'warehouse_name)
+      .orderBy('timestamp).rowsBetween(Window.currentRow, 1000)
+    val warehouse_name_gen_w = Window.partitionBy('organization_id, 'warehouse_id)
+      .orderBy('timestamp).rowsBetween(Window.currentRow, 1000)
+    val warehouse_id_gen = first('warehouse_id, true).over(warehouse_id_gen_w)
+    val warehouse_name_gen = first('warehouse_name, true).over(warehouse_name_gen_w)
+
+    val warehouseSummaryCols = auditBaseCols ++ Array[Column](
+      when('actionName === "createEndpoint" || 'actionName === "createWarehouse",
+        get_json_object($"response.result", "$.id"))
+        .when(('actionName =!= "createEndpoint" && 'id.isNull)
+          || ('actionName =!= "createWarehouse" && 'id.isNull), 'warehouseId)
```

Comment:
the most exclusive must come first in your case statement or this case will never be hit since the one above it will match first.

Reply:
please move this logic to a func like

`private def deriveWarehouseId(df: DataFrame) = when...when...otherwise...alias("warehouse_id")`

and then use this func as one of your selects. This will allow us to add unit tests in the future.

Reply:
done. created a dbsqlTransform.scala